File: interviews/A/VO/lca.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root, p, q):
        
        if root is None:
            return None 
        
        if root == p or root == q:
            return root 
        
        left_lca = self.lowestCommonAncestor(root.left, p, q)
        right_lca = self.lowestCommonAncestor(root.right, p, q)
        
        if left_lca and right_lca:
            
            return root 
        
        if left_lca:
            
            return left_lca 
        
        if right_lca:
            
            return right_lca 
        
        return None 
        

File: interviews/Twitch/JavaTwitch/LCA.java
import java.util.*; 

class TreeNode {

    int val; 
    TreeNode left;
    TreeNode right; 

    public TreeNode(int val) {

        this.val = val; 
        this.left = null; 
        this.right = null; 
    }
}

public class LCA {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        
        if (root == null) return null; 
        
        if (root == p || root == q) return root; 
        
        TreeNode leftLca = this.lowestCommonAncestor(root.left, p, q);
        TreeNode rightLca = this.lowestCommonAncestor(root.right, p, q);
        
        if (leftLca != null && rightLca != null) return root; 
        
        else if (leftLca != null) return leftLca; 
        
        else if (rightLca != null) return rightLca; 
        
        return null;
        
    }
}